File: android/src/main/java/com/oblador/keychain/KeychainModuleBuilder.java
package com.oblador.keychain;

import com.facebook.react.bridge.ReactApplicationContext;

public class KeychainModuleBuilder {
  public static final boolean DEFAULT_USE_WARM_UP = true;

  private ReactApplicationContext reactContext;
  private boolean useWarmUp = DEFAULT_USE_WARM_UP;

  public KeychainModuleBuilder withReactContext(ReactApplicationContext reactContext) {
    this.reactContext = reactContext;
    return this;
  }

  public KeychainModuleBuilder usingWarmUp() {
    useWarmUp = true;
    return this;
  }

  public KeychainModuleBuilder withoutWarmUp() {
    useWarmUp = false;
    return this;
  }

  public KeychainModule build() {
    validate();
    if (useWarmUp) {
      return KeychainModule.withWarming(reactContext);
    } else {
      return new KeychainModule(reactContext);
    }
  }

  private void validate() {
    if (reactContext == null) {
      throw new Error("React Context was not provided");
    }
  }
}

File: android/src/main/java/com/oblador/keychain/KeychainModuleBuilder.kt
package com.oblador.keychain

import com.facebook.react.bridge.ReactApplicationContext

class KeychainModuleBuilder {
    private var reactContext: ReactApplicationContext? = null
    private var useWarmUp = DEFAULT_USE_WARM_UP
    fun withReactContext(reactContext: ReactApplicationContext?): KeychainModuleBuilder {
        this.reactContext = reactContext
        return this
    }

    fun usingWarmUp(): KeychainModuleBuilder {
        useWarmUp = true
        return this
    }

    fun withoutWarmUp(): KeychainModuleBuilder {
        useWarmUp = false
        return this
    }

    fun build(): KeychainModule {
        validate()
        return if (useWarmUp) {
            KeychainModule.withWarming(reactContext!!)
        } else {
            KeychainModule(reactContext!!)
        }
    }

    private fun validate() {
        if (reactContext == null) {
            throw Error("React Context was not provided")
        }
    }

    companion object {
        const val DEFAULT_USE_WARM_UP = true
    }
}
